### PR TITLE
Add benches for 'propose' and 'ready' of RawNode

### PR DESCRIPTION
**Pull request by Fullstop000**
Add benches for `RawNode::propose` and `RawNode::ready` as they are two of the most essential API to upper layers.

Part of https://github.com/pingcap/raft-rs/issues/109

@Hoverbear @hicqu 

Signed-off-by: Fullstop000 <fullstop1005@gmail.com>

### CODE REVIEW

File: benches/suites/raw_node.rs [hunk 1]
```diff
@@ -20,3 +26,93 @@ pub fn bench_raw_node_new(c: &mut Criterion) {
...
+            node.raft.become_leader();
+            let mut total = Duration::from_nanos(0);
+            for _ in 0..iters {
+                // TODO: Maybe simulate more environments as input. For now, just preparing a raft node after stepping a proposal
+                node.propose(vec![], vec![]).expect("");
+                if node.has_ready() {
+                    let now = Instant::now();
+                    let ready = node.ready();
+                    total = total.add(now.elapsed());
+                    handle_ready(&mut node, ready);
```
Comment (BusyJay) [inline benches/suites/raw_node.rs]: Handle_ready usually depends on user usage, may not be proper counted in benchmark time.

Comment (Fullstop000) [reply benches/suites/raw_node.rs]: `handle_ready` is excluded from the benchmark as it's called after `total = total.add(now.elapsed())`.

**Comment by Fullstop000** [comment]
I refactored the bench for `RawNode::propose` by `Benchmark` which makes configuring measurement for single benchmark become possible.

@Hoverbear @BusyJay

File: benches/suites/raw_node.rs [hunk 2]
```diff
@@ -1,13 +1,18 @@
-use criterion::{Bencher, Criterion};
-use raft::{storage::MemStorage, Config, RawNode};
+use criterion::{Bencher, Benchmark, Criterion, Throughput};
```
Comment (Hoverbear) [inline benches/suites/raw_node.rs]: I note that Criterion 0.3.0 (the new version) has no `Benchmark` struct, so maybe the previous way was more forwards compatible? https://docs.rs/criterion/0.3.0/criterion/index.html

Comment (Fullstop000) [reply benches/suites/raw_node.rs]: Done. `BenchmarkGroup` works perfectly on this :)

File: benches/suites/raw_node.rs (hunk 1 above)
Comment (bheisler) [reply benches/suites/raw_node.rs]: Criterion.rs author here! Obviously I don't know much about your code, so feel free to ignore me, but doing `Instant::now()` and `Instant::elapsed()` for every iteration will introduce some nasty measurement overhead and skew your measurements. If you can, setting up a batch of nodes and calling `ready` on all of them using `iter_batched` would give more accurate measurements.

Comment (Fullstop000) [reply benches/suites/raw_node.rs]: Thank you for your guidance! @bheisler
`iter_batched` seems very suitable for the situations I met.

**Comment by BusyJay** [comment]
@hicqu @gengliqi PTAL

File: benches/suites/raw_node.rs (hunk 1 above)
Comment (Hoverbear) [reply benches/suites/raw_node.rs]: Thanks for your amazing library, @bheisler!

File: benches/suites/raw_node.rs [hunk 3]
```diff
@@ -20,3 +25,107 @@ pub fn bench_raw_node_new(c: &mut Criterion) {
...
+    group
+        // TODO: The proper measurement time could be affected by the system and machine.
+        .measurement_time(Duration::from_secs(60))
+        .bench_function("Default", |b: &mut Bencher| {
+            b.iter_batched(
+                || test_ready_raft_node(&logger),
+                |mut node| {
+                    let _ = node.ready();
+                },
+                // NOTICE: SmallInput accumulates (iters + 10 - 1) / 10 smaples per batch
```
Comment (Hoverbear) [inline benches/suites/raw_node.rs]:
```suggestion
                // NOTICE: SmallInput accumulates (iters + 10 - 1) / 10 samples per batch
```

File: benches/suites/raw_node.rs [hunk 4]
```diff
@@ -20,3 +25,107 @@ pub fn bench_raw_node_new(c: &mut Criterion) {
...
+            );
+    }
+}
+
+pub fn bench_raw_node_new_ready(c: &mut Criterion) {
+    let logger = crate::default_logger();
+    let mut group = c.benchmark_group("RawNode::ready");
+    group
+        // TODO: The proper measurement time could be affected by the system and machine.
+        .measurement_time(Duration::from_secs(60))
```
Comment (Hoverbear) [inline benches/suites/raw_node.rs]: 60 seems like a long time, can you help me understand the reasoning?

File: benches/suites/raw_node.rs [hunk 5]
```diff
@@ -20,3 +25,107 @@ pub fn bench_raw_node_new(c: &mut Criterion) {
...
+        .bench_function("Default", |b: &mut Bencher| {
+            b.iter_batched(
+                || test_ready_raft_node(&logger),
+                |mut node| {
+                    let _ = node.ready();
```
Comment (Hoverbear) [inline benches/suites/raw_node.rs]: From reading https://github.com/bheisler/criterion.rs/blob/master/book/src/faq.md#when-should-i-use-criterionblack_box I think we don't need to use it here. :)

File: benches/suites/raw_node.rs (hunk 4 above)
Comment (Fullstop000) [reply benches/suites/raw_node.rs]: If the measurement time isn’t long enough, the benchmark result could be inaccurate. Maybe using a 'lighter' node(e.g. with fewer committed entries and messages) ?

File: benches/suites/raw_node.rs (hunk 5 above)
Comment (Fullstop000) [reply benches/suites/raw_node.rs]: Do you mean the macro black_box? I haven’t used it in our benches.

Comment (Hoverbear) [reply benches/suites/raw_node.rs]: Yes I think the code is fine, I just was investigating for myself to make sure we didn't need to consider it. Good job. :)

File: benches/suites/raw_node.rs (hunk 4 above)
Comment (Hoverbear) [reply benches/suites/raw_node.rs]: Ok, I think this is a good reason.

File: benches/suites/raw_node.rs [hunk 6]
```diff
@@ -20,3 +25,107 @@ pub fn bench_raw_node_new(c: &mut Criterion) {
...
+        16 * KB,
+        128 * KB,
+        512 * KB,
+        KB * KB,
+    ];
+    let mut group = c.benchmark_group("RawNode::leader_propose");
+    for size in test_sets.drain(..) {
+        // Calculate measurement time in seconds according to the input size.
+        // The approximate time might not be the best but should work fine.
+        let mtime = if size < KB {
```
Comment (Hoverbear) [inline benches/suites/raw_node.rs]: Maybe it's better to just use `7` for everything?

Comment (Fullstop000) [reply benches/suites/raw_node.rs]: I'm worried that the total measurement time will be too long. We can still get enough analysing samples from the benchmark with tiny throughput by using shorter benching time.

**Comment by hicqu** [comment]
Seems it can't work with criterion `0.2`. In my environment it only works after I update criterion to 0.3. Could you check it @Fullstop000 ? If you also meet the same problem, please update the depencency in Cargo.toml.

**Comment by hicqu** [comment]
rest LGTM. Good job!

**Comment by Fullstop000** [comment]
Can I update `Criterion` to `0.3.0` directly? I'm worried that there might be some dependency concerns involved.
@brson @hicqu

**Comment by hicqu** [comment]
According to my test, it's ok to update criterion to 0.3.

**Comment by Hoverbear** [comment]
Seems the CLA is stuck but you have already signed it. xD

**Pull request merged**